studio/components/interfaces/Organization/OAuthApps/PublishAppModal.tsx

```diff
@@ -0,0 +1,314 @@
+    if (!isValidHttpUrl(values.website)) errors.website = 'Please provide a valid URL for your site'
+    return errors
+  }
+
+  const onSubmit = async (values: any, { setSubmitting }: any) => {
+    if (!slug) return console.error('Slug is required')
+
+    const redirect_uris = urls.filter((url) => url.value.length > 0).map((url) => url.value)
+    if (redirect_uris.length === 0) {
+      setSubmitting(false)
+      setErrors({ urls: 'Please provide at least one URL' })
+      return
+    } else {
+      setErrors({})
+    }
+
+    const { name, website } = values
+    const uploadedIconUrl =
+      iconFile !== undefined ? await uploadAttachment(slug, iconFile) : iconUrl
+
+    if (selectedApp === undefined) {
+      // Create application
+      try {
+        const res = await createOAuthApp({
+          slug,
+          name,
+          website,
+          redirect_uris,
+          icon: uploadedIconUrl,
+        })
+        ui.setNotification({
+          category: 'success',
+          message: `Successfully created OAuth app "${name}"!`,
+        })
+        onClose()
+        onCreateSuccess(res)
+      } catch (error: any) {
+        setSubmitting(false)
+        ui.setNotification({
+          category: 'error',
+          message: `Failed to create OAuth app: ${error.message}`,
+        })
+      }
+    } else {
+      // Update application
+      try {
+        await updateOAuthApp({
+          id: selectedApp.id,
+          slug,
+          name,
+          website,
+          redirect_uris,
+          icon: uploadedIconUrl === undefined ? null : uploadedIconUrl,
+        })
+        ui.setNotification({
+          category: 'success',
+          message: `Successfully updated OAuth app "${name}"!`,
+        })
+        onClose()
+      } catch (error: any) {
+        setSubmitting(false)
+        ui.setNotification({
+          category: 'error',
+          message: `Failed to update OAuth app: ${error.message}`,
+        })
+      }
+    }
+  }
+
+  return (
+    <Modal
+      hideFooter
+      visible={visible}
+      header={
+        selectedApp !== undefined ? 'Update OAutha application' : 'Publish a new OAuth application'
+      }
+      onCancel={() => onClose()}
+    >
+      <Form
+        validateOnBlur
+        initialValues={{ name: '', website: '' }}
+        validate={validate}
+        onSubmit={onSubmit}
+      >
+        {({ isSubmitting, resetForm }: { isSubmitting: boolean; resetForm: any }) => {
+          // [Joshen] although this "technically" is breaking the rules of React hooks
+          // it won't error because the hooks are always rendered in the same order
+          // eslint-disable-next-line react-hooks/rules-of-hooks
+          useEffect(() => {
+            if (visible && selectedApp !== undefined) {
+              const values = { name: selectedApp.name, website: selectedApp.website }
+              resetForm({ values, initialValues: values })
+            }
+          }, [visible, selectedApp])
```

Comment:
🥲

@MildTomato have we got some new form components in the works so we don't have to do this anymore?

Reply:
oh yes!

https://ui.shadcn.com/docs/components/form

Reply:
We might also move to Zod validation, but Yup() also still works with react-hook-form